app/models/middleware_server_performance.rb

```diff
@@ -0,0 +1,69 @@
+class MiddlewareServerPerformance < ActsAsArModel
+  set_columns_hash(
+    :id            => :string,
+    :name          => :string,
+    :resource_type => :string,
+    :feed          => :string,
+    :ems_ref       => :string,
+    :start_time    => :datetime,
+    :end_time      => :datetime
+  )
+
+  alias resource_id id
+  alias resource_name name
+
+  def self.build_results_for_report_middleware(options)
+    start_time, end_time, interval = parse_time_interval(options)
+    results = []
+    MiddlewareServer.find_each do |ms|
+      raw_stats = fetch_raw_stats(ms, start_time, end_time, interval)
+      if raw_stats.values[0]
+        columns = raw_stats.values[0].keys
+        columns.each { |column| set_columns_hash(column => :float) }
+        raw_stats.each { |timestamp, stats| results.push(parse_row(ms, timestamp, interval, stats)) }
+      end
+    end
+    [results, {}]
+  end
+
+  def self.parse_time_interval(options)
+    now = Time.now.utc
+    start_time = now - options[:start_offset]
+    end_time = now - options[:end_offset]
+    interval = case options[:interval]
+               when 'daily'   then 60 * 60 * 24
+               when 'hourly'  then 60 * 60
+               else end_time - start_time
+               end
+    [start_time, end_time, interval]
+  end
+
+  def self.fetch_raw_stats(middleware_server, start_time, end_time, interval)
+    raw_stats = {}
+    middleware_server.metrics_available.each do |metric|
+      middleware_server.collect_stats_metric(metric, start_time, end_time, interval).each do |stat|
+        timestamp = stat['start']
+        raw_stats[timestamp] = {} if raw_stats[timestamp].nil?
+        %w(min avg median max samples).each do |w|
```

Comment:
this array is part of the internal representation of the metrics raw data, which is processed at this level.
'samples' means the number of points for a bucket.
Once we introduce more report types we can gather and perhaps reorganize that if it makes sense.
For this one I think this is an implementation detail and it makes sense to maintain it at that level.